src/racer/nameres.rs

```diff
@@ -946,68 +946,49 @@ pub fn search_scope(start: Point, point: Point, src: Src,
     out.into_iter()
 }
 
-fn try_to_match_closure_definition(searchstr: &str, scope_src: &str, scope_src_pos: Point, filepath: &Path) -> Option<Match> {
+fn search_closure_args(searchstr: &str, scope_src: &str, scope_src_pos: Point,
+                       filepath: &Path, search_type: SearchType) -> Option<Vec<Match>> {
     if searchstr.is_empty() {
         return None;
     }
 
-    /// Search for a closure declaration (defined here as "stuff bracketed by | characters").
-    ///
-    /// The pipe characters are preserved by the capture as their presence for some reason
-    /// prevents bad matches on type annotations.
-    ///
-    /// TODO: properly look for closures by requiring it not be after an expression.
-    lazy_static! {
-        static ref CLOSURE_DECL: Regex = Regex::new(r"\|[^;]*?\|").unwrap();
-    }
-
     trace!("Closure definition match is looking for `{}` in {} characters", searchstr, scope_src.len());
 
-    if let Some(cap) = CLOSURE_DECL.captures(scope_src) {
-        /// This regex looks for the passed-in ident surrounded by non-ident characters
-        /// on the left side of a colon (if one is present). It takes as input the declaration
-        /// of a closure's arguments.
-        let ident_in_closure = Regex::new(&format!(r"(?x:
-            \|
-            (?:(?s).*,[^a-zA-Z0-9]*|[^a-zA-Z0-9]*)? # discard things before the identifier
-            (?P<definition>{})
-            (?:[^a-zA-Z0-9\|]+[^\|;]*)?
-            \|
-        )", searchstr)).unwrap();
-
-        let decl = cap.get(0).unwrap();
-
-        trace!("Found a closure declaration {}..{}, about to search for `{}` in {}", 
-            decl.start(), 
-            decl.end(), 
-            searchstr, 
-            decl.as_str().trim()
-        );
-
-        if let Some(ident_match) = ident_in_closure.captures(decl.as_str()) {
-
-            let def = ident_match.name("definition").unwrap();
-    
-            trace!("Closure definition matched by {}..{}: {:?}", def.start(), def.end(), &cap[0]);
-
-            Some(Match {
-                matchstr: searchstr.to_owned(),
-                filepath: filepath.to_path_buf(),
-                point: def.start() + scope_src_pos,
-                coords: None,
-                local: true,
-                mtype: FnArg,
-                contextstr: cap.get(0).unwrap().as_str().to_owned(),
-                generic_args: Vec::new(),
-                generic_types: Vec::new(),
-                docs: String::new(),
-            })
-        } else {
-            None
+    if let Some((left_pipe, _, pipe_scope)) = closure_valid_arg_scope(scope_src) {
```

Comment:
Can you please add a `trace!` statement here which reports that the function has found a closure?